### DISCUSSION TITLE
How to get the node value of coupled variable in current element for 1D mesh?

### DISCUSSION THREAD

**Discussion by YuhangNIU** [idaholab/moose#19673] [Q&A General] [answered]
Hi all,

I know the node value of mainly variable in the current element (in ADKernel) can be defined as "_u_nodal(_var.dofValues())" and can be used as _u_nodal[0] in 1D mesh. But how to define and get the node value of the coupled variable in the current element?
 
Could anyone kindly teach me?

Best regards,
Yuhang

**Comment by GiudGiud** [comment, marked as answer]
Hello

Are you talking about a coupled auxiliary variable?
You cant define directly the nodal (or any) value of a main (non-linear) variable, these have to be solved for. 

Guillaume

**Reply by YuhangNIU** [reply]
Thank you for your reply!

I have solved this problem in another way. In this problem, we could define the coupled variable as **_coupled_var_nodal(getVar("coupled_var", 0)->dofValues())**. So the node value of the coupled variable in the current element can be expressed as **_coupled_var_nodal[0]** or **_coupled_var_nodal[1]** in 1D mesh.

Much appreciate!

Yuhang

**Reply by GiudGiud** [reply]
sounds good. 
You can indeed access a coupled variable dof that way. The preferred way is to use _coupled_var_nodal(coupledValue("coupled_var", 0))    **(adCoupledValue for ADKernels)**

Please see the function definition in interface/Coupleable.C and see if you want to use that instead:
```
const VariableValue &
Coupleable::coupledValue(const std::string & var_name, unsigned int comp) const
{
  const auto * const var = getVarHelper<MooseVariableField<Real>>(var_name, comp);
  if (!var)
    return *getDefaultValue(var_name, comp);
  checkFuncType(var_name, VarType::Ignore, FuncAge::Curr);

  if (!_coupleable_neighbor)
  {
    if (_c_nodal)
      return (_c_is_implicit) ? var->dofValues() : var->dofValuesOld();
    else
      return (_c_is_implicit) ? var->sln() : var->slnOld();
  }
  else
  {
    if (_c_nodal)
      return (_c_is_implicit) ? var->dofValuesNeighbor() : var->dofValuesOldNeighbor();
    else
      return (_c_is_implicit) ? var->slnNeighbor() : var->slnOldNeighbor();
  }
}
```
_coupled_var_nodal is of the `VariableValue` type.  (ADVariableValue for AD)
It can be accessed at a quadrature point with `[_qp]`

**Reply by YuhangNIU** [reply]
Thank you! It looks better. I will have a try in this way.

Best regards,
Yuhang
